### DISCUSSION TITLE
2d_segmentation: Where is "best_metric_model_segmentation2d_array.pth" in load.torch?

### DISCUSSION THREAD

**Discussion by GamblerEd** [Project-MONAI/MONAI#3851] [Q&A] [answered]
Hello everyone,

I am an El. Engineering student from TUM and as you can see, I have just made an 
account of mine on GitHub and I'm using the platform for the first time really.
Please forgive me if my question is not formulated properly or if the question is silly.

As part of a project I have to do in my Python Data Engineering course, I got proposed
to do some experiments on the MONAI project. The particular subject of my project is
image segmentation using Unet architecture, so I am basically keen on looking at the codes
in MONAI/tutorials/2d_segmentation/torch.

I am equipped with my VSC editor and am using an OpenVPN GUI SSH on to have access on 
uni computers/GPUs so that I can use these GPUs for working with the data in MONAI. 
I should mention that I have had no experience in using virtual PCs on VSC for programming 
(through SLURM), and I have neither had any experience with GitHub, so please be patient, 
I will do my best to understand how to work with the platform and your codes.

So far, I have cloned the repositories Project-MONAI/MONAI and Project-MONAI/tutorials,
I have further read the README.md file of the tutorials repository and have executed the 
commands listed in ### 1. Requirements.

Here is the issue I am encountering:
When I try to run the code unet_evaluation_array.py from my "[<mypers_uni_id>@TUEINAN-LOGIN1 ~]$"-terminal,
the execution fails because I get the FileNotFoundError, here the whole error log:

```
MONAI version: 0.9.dev2208
Numpy version: 1.21.5
Pytorch version: 1.10.2+cu102
MONAI flags: HAS_EXT = False, USE_COMPILED = False
MONAI rev id: cb1f52055bd35ca6c2d7ccedfd997f6a0683c01f
MONAI __file__: /home/ge25ron/.local/lib/python3.7/site-packages/monai/__init__.py

Optional dependencies:
Pytorch Ignite version: 0.4.8
Nibabel version: 3.2.2
scikit-image version: 0.19.2
Pillow version: 9.0.1
Tensorboard version: 2.8.0
gdown version: 4.3.1
TorchVision version: 0.11.3+cu102
tqdm version: 4.62.3
lmdb version: 1.3.0
psutil version: 5.9.0
pandas version: 1.3.5
einops version: 0.4.0
transformers version: 4.16.2
mlflow version: 1.23.1

For details about installing the optional dependencies, please visit:
    https://docs.monai.io/en/latest/installation.html#installing-the-recommended-dependencies

generating synthetic data to /tmp/tmpb9emz3e9 (this may take a while)
Traceback (most recent call last):
  File "/home/ge25ron/tutorials/2d_segmentation/torch/unet_evaluation_array.py", line 85, in <module>
    main(tempdir)
  File "/home/ge25ron/tutorials/2d_segmentation/torch/unet_evaluation_array.py", line 62, in main
    model.load_state_dict(torch.load("best_metric_model_segmentation2d_array.pth"))
  File "/home/ge25ron/.local/lib/python3.7/site-packages/torch/serialization.py", line 594, in load
    with _open_file_like(f, 'rb') as opened_file:
  File "/home/ge25ron/.local/lib/python3.7/site-packages/torch/serialization.py", line 230, in _open_file_like
    return _open_file(name_or_buffer, mode)
  File "/home/ge25ron/.local/lib/python3.7/site-packages/torch/serialization.py", line 211, in __init__
    super(_open_file, self).__init__(open(name, mode))

FileNotFoundError: [Errno 2] No such file or directory: 'best_metric_model_segmentation2d_array.pth'
```

So basically the function `torch.load("best_metric_model_segmentation2d_array.pth")` cannot be executed because
"best_metric_model_segmentation2d_array.pth" is nowhere to be found.

Has this maybe to do with the point brought forth in #### Data (README.md), which states that some notebooks 
will require additional data, and that they should be downloaded by running the [runexamples.sh](./runexamples.sh) script.

If that is the case, how exactly am I supposed to run this runexamples.sh-script? Since I have to use the virtual computers,
I'd have to set up a task using SLURM commands to actually run runexamples.sh .
Like I said before, I have never done this, and the commands on how to run tasks seem to be very complicated and 
incomprehensible to me. 
Is there anybody here who could help me on this particular issue? I don't think I can solve this alone at the moment.
Did I even begin my work on MONAI properly, are there even other repositories or packages/dependencies that I 
should download too?

I hope someone will very soon give me some help, I'd appreciate it very much and I'm very thankful for the help in advance!

I am very interested in getting to know project MONAI and I'm sure I'll have much fun working with it and spending my time
with the community here!

Cheers all

**Comment by Nic-Ma** [comment, marked as answer]
Hi @GamblerEd ,

Thanks for your interest here.
The `best_metric_model_segmentation2d_array.pth` is trained by another example program, so please try to run the train example in the same folder first.
For more questions, I think @yiheng-wang-nv can help address them.

Thanks.

**Reply by GamblerEd** [reply]
Hello @Nic-Ma ,
Please forgive me for replying too late, I was busy with other work.
Could you tell me which program file it is where `best_metric_model_segmentation2d_array.pth` 
is trained exactly, is it one of those other 3 .py-files of the _2d_segmentation/torch_ folder?
How can I reach @yiheng-wang-nv (sry but I'm new), does he see my post when I mention/@-tag him here?

Thank you very much for your answer!

Cheers.

EDIT:
I have now found the command `torch.save(model.state_dict(), "best_metric_model_segmentation2d_array.pth")` in `unet_training_array.py` , when I try execute it, I get this error message:
`AttributeError: module 'distutils' has no attribute 'version'`
What does this mean? @yiheng-wang-nv 

EDIT:
I have fixed this problem, I have figured it out, it's okay!

**Reply by Nic-Ma** [reply]
Cool, glad to see your progress!

Thanks.
